File: TripKitUI/views/map annotations/TKUICircleAnnotationView.m
//
//  TKUICircleAnnotationView.m
//  TripKit
//
//  Created by Adrian Schönig on 16/06/11.
//  Copyright 2011 SkedGo. All rights reserved.
//

#import "TKUICircleAnnotationView.h"

#import "UIColor+Variations.h"

#define kBHCircleSize   12.0
#define kBHSmallFactor    .8 // percentage

@implementation TKUICircleAnnotationView

- (id)initWithAnnotation:(id <MKAnnotation>)annotation 
               drawLarge:(BOOL)large 
         reuseIdentifier:(NSString *)reuseIdentifier
{
  self = [super initWithAnnotation:annotation reuseIdentifier:reuseIdentifier];
  if (self) {
    CGFloat radius = kBHCircleSize;
    if (! large) radius *= kBHSmallFactor;

    self.isLarge = large;
    self.frame = CGRectMake(0.0f, 0.0f, radius, radius);
    self.opaque = YES;
    self.backgroundColor = [UIColor clearColor];
		self.isFaded = NO;
  }

  return self;
}

- (void)drawRect:(CGRect)rect
{
#pragma unused(rect) // we are inefficient :(

#define LINE_WIDTH 1.5f
	CGFloat lineWidth = LINE_WIDTH * (self.isLarge ? 1 : 0.8f);
	
	// determine the colors
	UIColor *lineColor = self.circleColor;
  if (nil == lineColor)
    lineColor = [UIColor blackColor];
	
	UIColor *borderColor = [lineColor darkerColorByPercentage:0.75];
  
  CGContextRef context = UIGraphicsGetCurrentContext();
  
  // draw circle
  CGFloat lineOffset = lineWidth / 2.0f;
  CGRect  circleRect = CGRectMake(lineOffset, lineOffset, self.bounds.size.width - lineWidth, self.bounds.size.width - lineWidth);

	if (self.isFaded) {
		CGContextSetFillColorWithColor(context, lineColor.CGColor);
	} else {
		CGContextSetFillColorWithColor(context, [UIColor whiteColor].CGColor);
	}
  CGContextFillEllipseInRect(context, circleRect);
	CGContextSetLineWidth(context, lineWidth);
	CGContextSetStrokeColorWithColor(context, borderColor.CGColor);
  CGContextStrokeEllipseInRect(context, circleRect);
}
@end

File: Sources/TripKitUI/views/map annotations/TKUICircleAnnotationView.swift
//
//  TKUICircleAnnotationView.swift
//  TripKitUI-iOS
//
//  Created by Adrian Schönig on 12.12.18.
//  Copyright © 2018 SkedGo Pty Ltd. All rights reserved.
//

import Foundation
import MapKit

public protocol TKUICircleDisplayable: MKAnnotation {
  var circleColor: UIColor { get }
  var isTravelled: Bool { get }
  var asLarge: Bool { get }
}

open class TKUICircleAnnotationView: MKAnnotationView {
  private enum Constants {
    static let circleSize: CGFloat   = 12.0
    static let smallFactor: CGFloat  = 0.8
    static let lineWidth: CGFloat    = 1.5
  }
  
  public let isLarge: Bool
  
  public var circleColor: UIColor? = nil
  public var borderColor: UIColor? = nil
  public var isFaded: Bool = false
  
  public init(annotation: MKAnnotation?, drawLarge: Bool, reuseIdentifier: String?) {
    self.isLarge = drawLarge
    
    super.init(annotation: annotation, reuseIdentifier: reuseIdentifier)
    
    let radius = Constants.circleSize * (drawLarge ? 1 : Constants.smallFactor)
    frame.size = CGSize(width: radius, height: radius)
    backgroundColor = .clear
  }
  
  public required init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  public override func draw(_ rect: CGRect) {
    guard let context = UIGraphicsGetCurrentContext() else { return }
    
    let fillColor = self.circleColor ?? .black
    let borderColor = self.borderColor ?? fillColor.darkerColor(percentage: 0.75)
    let lineWidth = Constants.lineWidth * (isLarge ? 1 : Constants.smallFactor)
    let lineOffset = lineWidth / 2
    
    let circleRect = CGRect(x: lineOffset, y: lineOffset, width: bounds.width - lineWidth, height: bounds.height - lineWidth)
    if isFaded || circleColor != nil {
      context.setFillColor(fillColor.cgColor)
    } else {
      context.setFillColor(UIColor.tkBackgroundNotClear.cgColor)
    }
    context.fillEllipse(in: circleRect)
    
    context.setLineWidth(lineWidth)
    context.setStrokeColor(borderColor.cgColor)
    context.strokeEllipse(in: circleRect)
  }
  
}
